contracts/external/proxies/LiquidatorProxyV1ForSoloMargin.sol

```diff
@@ -0,0 +1,433 @@
+            "Liquid account not liquidatable",
+            liquidSupplyValue.value,
+            liquidBorrowValue.value
+        );
+
+        uint256 numMarkets = SOLO_MARGIN.getNumMarkets();
+        // for each owedMarket
+        for (uint256 owedMarket = 0; owedMarket < numMarkets; owedMarket++) {
+            // for each heldMarket
+            for (uint256 heldMarket = 0; heldMarket < numMarkets; heldMarket++) {
+                // cannot use the same market
+                if (heldMarket == owedMarket) {
+                    continue;
+                }
+
+                // cannot liquidate non-negative markets
+                if (!SOLO_MARGIN.getAccountPar(liquidAccount, owedMarket).isNegative()) {
+                    break;
+                }
+
+                // cannot use non-positive markets as collateral
+                if (!SOLO_MARGIN.getAccountPar(liquidAccount, heldMarket).isPositive()) {
+                    continue;
+                }
+
+                // get all relevant values
+                Cache memory cache = initializeCache(
+                    fromAccount,
+                    heldMarket,
+                    owedMarket,
+                    minLiquidatorRatio
+                );
+
+                // do the liquidation
+                liquidateInternal(
+                    fromAccount,
+                    liquidAccount,
+                    cache,
+                    heldMarket,
+                    owedMarket
+                );
+            }
+        }
+    }
+
+    function liquidateInternal(
+        Account.Info memory fromAccount,
+        Account.Info memory liquidAccount,
+        Cache memory cache,
+        uint256 heldMarket,
+        uint256 owedMarket
+    )
+        private
+    {
+        // get it to the state where owed is negative and held is positive (in the liquidator acct)
+        liquidateUntilNegOwedAndPosHeld(cache);
+
+        // figure out where to stop
+        figureOutWhereToStop(cache);
+
+        // if nothing to liquidate, don't
+        if (cache.liquidated == 0) {
+            return;
+        }
+
+        // create liquidate action
+        Actions.ActionArgs[] memory actions = new Actions.ActionArgs[](1);
+        actions[0] = Actions.ActionArgs({
+            actionType: Actions.ActionType.Liquidate,
+            accountId: 0,
+            amount: Types.AssetAmount({
+                sign: true,
+                denomination: Types.AssetDenomination.Wei,
+                ref: Types.AssetReference.Delta,
+                value: cache.liquidated
+            }),
+            primaryMarketId: owedMarket,
+            secondaryMarketId: heldMarket,
+            otherAddress: address(0),
+            otherAccountId: 1,
+            data: new bytes(0)
+        });
+
+        // create accounts
+        Account.Info[] memory accounts = new Account.Info[](2);
+        accounts[0] = fromAccount;
+        accounts[1] = liquidAccount;
+
+        // execute the liquidations
```

Comment:
what's your thought on doing the liquidations in separate operations vs the same operations?

It seems like same operations would be less gas, and kind of conceptually makes sense.

Or is the problem you need the account balances to be updated by the time you consider the next liquidation?

Reply:
Yeah it's just way easier to do it this way TBH